pep-0636.rst

```diff
@@ -0,0 +1,382 @@
+
+Alternate patterns
+------------------
+
+Going back to the adventure game example, you may find that you'd like to have several
+patterns resulting in the same outcome. For example, you might want the commands 
+``north`` and ``go north`` be equivalent. You may also desire to have aliases for 
+``get X``, ``pick up X`` and ``pick X up`` for any X. 
+
+The ``|`` symbol in patterns combines them as alternatives. You could for example write::
+
+   match command.split():
+       ... # Other case clauses
+       case ["north"] | ["go", "north"]:
+           current_room = current_room.neighbor("north")
+       case ["get", obj] | ["pick", "up", obj] | ["pick", obj, "up"]:
+           ... # Code for picking up the given object
+
+This is called an **or pattern** and will produce the expected result. Patterns are
+attempted from left to right; this may be relevant to know what is bound if more than
+one alternative matches. An important restriction when writing or patterns is that all
+alternatives should bind the same variables. So a pattern ``[1, x] | [2, y]`` is not
+allowed because it would make unclear which variable would be bound after a successful
+match. ``[1, x] | [2, x]`` is perfectly fine and will always bind ``x`` if successful.
+
+
+Capturing matched sub-patterns
+------------------------------
+
+An older version of our "go" command was written with a ``["go", direction]`` pattern.
+The change we did in our last version using the pattern ``["north"] | ["go", "north"]``
+has some benefits but also some drawbacks in comparison: the latest version allows the
+alias, but also has the direction hardcoded, which will force us to actually have
+separate patterns for north/south/east/west. This leads to some code duplication, but at
+the same time we get better input validation, and we will not be getting into that
+branch if the command entered by the user is ``"go figure!"`` instead of an direction.
+
+We could try to get the best of both worlds doing the following (I'll omit the aliased
+version without "go" for brevity)::
+
+   match command.split():
+       case ["go", ("north" | "south" | "east" | "west")]:
+           current_room = current_room.neighbor(...)
+           # how do I know which direction to go?
+
+This code is a single branch, and it verifies that the word after "go" is really a
+direction. But the code moving the player around needs to know which one was chosen and
+has no way to do so. What we need is a pattern that behaves like the or pattern but at
+the same time does a capture. We can do so with a **walrus pattern**::
+
+   match command.split():
+       case ["go", direction := ("north" | "south" | "east" | "west")]:
+           current_room = current_room.neighbor(direction)
+
+The walrus pattern (named like that because the ``:=`` operator looks like a sideways
+walrus) matches whatever pattern is on its right hand side, but also binds the value to
+a name.
+
+Conditional pattern matching
```

Comment:
Maybe "Adding a condition"? Somehow this heading totally misdirected my mind for a few seconds.

Reply:
Rephrased as "adding condition to patterns". While I write this I'm thinking that "adding condition to cases" may be more accurate although it sounds a bit worse.